src/java/org/apache/cassandra/service/accord/AccordSafeState.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.accord;
+
+import java.util.function.Function;
+
+import accord.impl.SafeState;
+import accord.utils.async.AsyncChain;
+import accord.utils.async.AsyncResults;
+import org.apache.cassandra.service.accord.AccordLoadingState.LoadingState;
+
+public interface AccordSafeState<K, V> extends SafeState<V>
+{
+    void set(V update);
+    V original();
+    void preExecute();
+    void postExecute();
+    AccordLoadingState<K, V> global();
+
+    default boolean hasUpdate()
+    {
+        return original() != current();
+    }
+
+    default void revert()
+    {
+        set(original());
+    }
+
+    default K key()
+    {
+        return global().key();
+    }
+
+    default LoadingState loadingState()
+    {
+        return global().state();
+    }
+
+    default AsyncResults.RunnableResult<V> load(Function<K, V> loadFunction)
+    {
+        return global().load(loadFunction);
+    }
+
+    default AsyncChain<?> listen()
+    {
+        return global().listen();
+    }
+
+    default Throwable failure()
+    {
+        return global().failure();
+    }
+
+    default void checkNotInvalidated()
+    {
+        if (invalidated())
+            throw new IllegalStateException("Cannot access invalidated " + this);
```

Comment:
fixing the toString we get

```
ERROR [CommandStore[12]:1] 2023-03-02 18:00:33,137 AsyncOperation.java:274 - Operation AsyncOperation{SAVING}-0x7bd3b9c2 failed
java.lang.IllegalStateException: Cannot access invalidated AccordSafeCommand{invalidated=true, global=Node{LOADED, key=[1,1677808833023004,2,2130706433], references=1}@2de5a0d2, original=null, current=Command@1167976198{[1,1677808833023004,2,213
0706433]:PreAccepted}}
        at org.apache.cassandra.service.accord.AccordSafeState.checkNotInvalidated(AccordSafeState.java:74)
        at org.apache.cassandra.service.accord.AccordSafeCommand.original(AccordSafeCommand.java:94)
        at org.apache.cassandra.service.accord.AccordSafeCommand.original(AccordSafeCommand.java:29)
        at org.apache.cassandra.service.accord.AccordSafeState.hasUpdate(AccordSafeState.java:38)
        at org.apache.cassandra.service.accord.async.AsyncWriter.lambda$assembleWrites$1(AsyncWriter.java:85)
        at java.base/java.util.HashMap.forEach(HashMap.java:1337)
        at org.apache.cassandra.service.accord.async.AsyncWriter.assembleWrites(AsyncWriter.java:84)
        at org.apache.cassandra.service.accord.async.AsyncWriter.maybeDispatchWrites(AsyncWriter.java:116)
        at org.apache.cassandra.service.accord.async.AsyncWriter.save(AsyncWriter.java:155)
        at org.apache.cassandra.service.accord.async.AsyncOperation.runInternal(AsyncOperation.java:235)
        at org.apache.cassandra.service.accord.async.AsyncOperation.run(AsyncOperation.java:270)
        at org.apache.cassandra.service.accord.async.AsyncOperation.callback(AsyncOperation.java:158)
```

Reply:
```ERROR [CommandStore[5]:1] 2023-03-02 18:04:41,707 JVMStabilityInspector.java:68 - Exception in thread Thread[CommandStore[5]:1,5,CommandStore[5]]
java.lang.IllegalStateException: Cannot access invalidated AccordSafeCommand{invalidated=true, global=Node{LOADED, key=[1,1677809078513005,2,2130706433], references=3}@36b1ca9e, original=Command@1483650346{[1,1677809078513005,2,2130706433]:PreAc
cepted}, current=Command@1483650346{[1,1677809078513005,2,2130706433]:PreAccepted}}
        at org.apache.cassandra.service.accord.AccordSafeState.checkNotInvalidated(AccordSafeState.java:74)
        at org.apache.cassandra.service.accord.AccordSafeCommand.current(AccordSafeCommand.java:80)
        at org.apache.cassandra.service.accord.AccordSafeCommand.current(AccordSafeCommand.java:29)
        at org.apache.cassandra.service.accord.AccordSafeState.hasUpdate(AccordSafeState.java:38)
        at org.apache.cassandra.service.accord.async.AsyncWriter.lambda$assembleWrites$1(AsyncWriter.java:85)
        at java.base/java.util.HashMap.forEach(HashMap.java:1337)
        at org.apache.cassandra.service.accord.async.AsyncWriter.assembleWrites(AsyncWriter.java:84)
        at org.apache.cassandra.service.accord.async.AsyncWriter.maybeDispatchWrites(AsyncWriter.java:116)
        at org.apache.cassandra.service.accord.async.AsyncWriter.save(AsyncWriter.java:155)
        at org.apache.cassandra.service.accord.async.AsyncOperation.runInternal(AsyncOperation.java:235)
        at org.apache.cassandra.service.accord.async.AsyncOperation.run(AsyncOperation.java:270)```

Reply:
this makes sense due to the safe commands being global state, so invalidating after the first touch, we do that in ```accord.impl.AbstractSafeCommandStore#complete```

Reply:
`org.apache.cassandra.service.accord.async.AsyncOperation#runInternal` calls

```
commandStore.completeOperation(safeStore, context.commands, context.commandsForKeys);
```

which makes it unsafe for `writer` to have access to `org.apache.cassandra.service.accord.async.AsyncOperation.Context#commands`

So, for this to work we need to do `completeOperation` *after* the write, or make the writer context not have access.